paramiko/transport.py

```diff
@@ -1791,6 +2760,8 @@ def run(self):
                             self._log(ERROR, 'Channel request for unknown channel %d' % chanid)
                             self.active = False
                             self.packetizer.close()
+                        # This line is key. With it, the channel can go out of scope and shut everything down
+                        chan = None
```

Comment:
(when I was describing why various strategies would not work in the previous PR, I was not considering use of weak refs, but used as they are here they seem somewhat reasonable)